### ISSUE TITLE
some implementation details double check and a few questinos

### ISSUE THREAD

**Issue by RichardChangCA** [ThoroughImages/CAMEL#10]
Hello, Thanks for your responses several years ago. I have implemented your paper these days, but I also have some questions about the implementation details.

1. In cMIL training part, I know you use 4 single GPUs(1 image for 1 GPU) to average their losses. Since I only have one GPU with 12GB memory, I just shrink the original image to the half and use the batch size of 4 to simulate your training step. I wanna ask how to arrange images into the batch, I use 2 positive images and 2 negative images for one batch, is this same as yours?
2. In Label Enrichment Part, Do you split images into training, validation and testing dataset and use validation dataset to find the convergence condition? Or just use training curve to get the lowest loss? I mean do you consider overfitting in cMIL training part? or overfitting is not necessary because we only need to generate pesudo labels of all images.
3. How to evaluate the performance of cMIL training? I use accuracy, precision, recall, confusion matrix to evaluate it. If the prediction value of only one patch of the whole image is larger than 0.5(due to sigmoid activation function in the last layer), I recognize it as positive. In other words, prediction values of all patches are smaller than 0.5 and then I regard this image as negative. However, in this metric, my result is so bad, which is around 50% accuracy.

Since I plan to use your work as a benchmark in my thesis paper, your comments are important for me. Looking forward to your reply.

Best,
Lingfeng

**Comment by RichardChangCA** [comment]
In addition, if I use 2 positive images and 2 negative images for one batch with max-max function to train the model, the model will tend to predict all samples as positive, where loss can decrease during training, since the loss between positive_label and max_output is smaller than the loss between negative_label and max_output. However, this loss decreasing path is a wrong direction. :(

**Comment by thuxugang** [comment]
Hi, thanks for your insterest~
1. We didn't balance positive and negative images at batch-level (namely in each step). We only performed the data balance at image-level (namely the whole traning set).
2. We split images into training, validation and testing dataset in traditional way.
3. The threshold we used is 0.5.
I am not sure I understand the problem you encountered correctly, but here are some advice:
1. The model after training using max-max critersion will have lower recall (~56.9) and higher precision (~95) on the test set. The model after training using max-min critersion will have higher recall (~82) and lower precision (~75.2). If you can't achieve the similar accuracies, there may be some problems in the patch selection procedure.
2. Be sure you save the right patch in each image, you can check your max-max and max-min models on your selected data, the performance should be same as that on the test set.
3. Discard the patches derived from the NC images to relieve the false-positive issue.

**Comment by RichardChangCA** [comment]
Thanks for your help!

I got a decent result now. Have to tried more than one batch size? I think in this model, if batch size > 1, the model cannot learn anything, maybe the batch normalization is not suitable for patches of different original large images. I reduce the batch size from 2(or 4, just shrinking images into smaller ones) to 1, the result is decent now, although I used gradient accumulation.

In addition, I think training cMIL part is time-consuming, I only trained 10 epochs to get the decent result, but it costs 10 hrs. Could I ask you how much epochs you trained the cMIL part, please? 1 Epoch is 1 iteration of all training images.

Best,

**Comment by thuxugang** [comment]
We only tried batch=1 since we had multiple gpus
. But I think batch>1 should achieve similar results. 
Yes, it will take some times to get reasonable results, but usually within one epoch. That's why losses from multiple gpus are crucial for training. Training takes around 20 epoches, 4 gpus in 2 days.
